subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/report/DependencyVerificationReportWriter.java

```diff
@@ -0,0 +1,240 @@
+    private static final Comparator<Map.Entry<ModuleComponentArtifactIdentifier, Collection<RepositoryAwareVerificationFailure>>> BY_MODULE_ID = Comparator.comparing(e -> e.getKey().getDisplayName());
+
+    private final Path gradleUserHome;
+    private final AbstractTextDependencyVerificationReportRenderer summaryRenderer;
+    private final HtmlDependencyVerificationReportRenderer htmlRenderer;
+
+    public DependencyVerificationReportWriter(Path gradleUserHome,
+                                              DocumentationRegistry documentationRegistry,
+                                              File verificationFile,
+                                              List<String> writeFlags,
+                                              File htmlReportOutputDirectory,
+                                              boolean verboseConsoleReport) {
+        this.gradleUserHome = gradleUserHome;
+        this.summaryRenderer = createConsoleRenderer(gradleUserHome, documentationRegistry, verboseConsoleReport);
+        this.htmlRenderer = new HtmlDependencyVerificationReportRenderer(documentationRegistry, verificationFile, writeFlags, htmlReportOutputDirectory);
+    }
+
+    public AbstractTextDependencyVerificationReportRenderer createConsoleRenderer(Path gradleUserHome, DocumentationRegistry documentationRegistry, boolean verboseConsoleReport) {
+        if (verboseConsoleReport) {
+            return new TextDependencyVerificationReportRenderer(gradleUserHome, documentationRegistry);
+        }
+        return new SimpleTextDependencyVerificationReportRenderer(gradleUserHome, documentationRegistry);
+    }
+
+    public Report generateReport(String displayName,
+                                 Multimap<ModuleComponentArtifactIdentifier, RepositoryAwareVerificationFailure> failuresByArtifact) {
+        // We need at least one fatal failure: if it's only "warnings" we don't care
+        // but of there's a fatal failure AND a warning we want to show both
+        doRender(displayName, failuresByArtifact, summaryRenderer);
+        doRender(displayName, failuresByArtifact, htmlRenderer);
+        File htmlReport = htmlRenderer.writeReport();
+        return new Report(summaryRenderer.render(), htmlReport);
+    }
+
+    public void doRender(String displayName, Multimap<ModuleComponentArtifactIdentifier, RepositoryAwareVerificationFailure> failuresByArtifact, DependencyVerificationReportRenderer renderer) {
+        ReportState reportState = new ReportState();
+        renderer.title(displayName);
+        renderer.withErrors(() -> {
+            // Sorting entries so that error messages are always displayed in a reproducible order
+            failuresByArtifact.asMap()
+                .entrySet()
+                .stream()
+                .sorted(DELETED_LAST.thenComparing(MISSING_LAST).thenComparing(BY_MODULE_ID))
+                .forEachOrdered(entry -> {
```

Comment:
They can't really because it would break the "delegation" model of the writer chain. Each needs its own instance. I can refactor this more but not sure it would provide much value.